### DISCUSSION TITLE
784

### DISCUSSION THREAD

**Discussion by giscus** [jojoldu/blog-comments#955] [Comments]
# 784

RDBMS를 사용하다보면 Nested Loop Join 으로 인해 성능 저하를 겪는 경우가 많다.일부 RDBMS는 특정 버전에 따라 Nested Loop Join만 지원되는 경우도 있다.다만, 요즘의 RDBMS는 대부분 Hash Join등 여러 Join 형태를 지원하고 있고 이를 통해 성능 개선이 가능하다.MySQL도 8.0.18 부터 Hash Join을 지원한다.실제 사례로 성능 개선을 진행해보자.1. 문제 쿼리아래와 같이 여러 Join을 진행하는 실제 쿼리가 있다.SELECT i.id, i.titles[?] AS title, i.icon_url FROM institutions i     INNER JOIN interested_corporations ic ON i.id = ic.institution_i..

https://jojoldu.tistory.com/784

**Comment by cremazer** [comment]
문제의 인식과 그에 따른 해결책을 제시하여 진행하는 과정에 대해 이해하기 쉽게 설명해 주셔서 도움이 되었습니다. 공유해주셔서 고맙습니다. :+1:

**Comment by born-developer** [comment]
본문에 두가지 오류가 있는 것 같습니다.

1. 기존 쿼리와 개선 쿼리의 내용이 다릅니다. 개선 쿼리에서는 group by 가 1번이 아니라 2번 적용 되었고, 이 때문에 성능이 달라진 것으로 보입니다.

2. "조인 전에 데이터 양을 줄여서 조인 작업을 효율적으로 만든다." 이건 sub-query, CTE 등과 무관합니다. 기존 쿼리처럼 where 를 뒷부분에 쓰셔도  최적화 과정 중 predicate pushdown 이 일어나 join 전에 필터링이 먼저 일어나는건 마찬가지입니다.

말씀 드리고 싶은 점은 SQL 작성시 선언적으로 원하는것(what)만 적으면 되지, 실행 과정에 영향을 주기 위해 (흔히 얘기하는) 쿼리-튜닝(how)하는 것은 거의 의미가 없거나 최적화 엔진을 제약해 더 느릴수 있다는 점입니다. 사람이 생각할 수 있는 수준의 기초적인 최적화는 이미 오래전에 실행계획 엔진에 적용되어 있습니다.

**Reply by jojoldu** [reply]
오 네네 동일한 조회 결과를 서브쿼리 안에서 내기 위해 하다보니 Group by를 2번 사용했는데,
말씀하신대로 글의 목적을 달성하기 위해 호도한 부분이 있었네요 ^^;

사람이 생각할 수 있는 수준의 최적화는 엔진에서 당연히 해주는 것이지만,
그럼에도 사람이 해줄 수 있는 부분은 있다고 생각합니다 :) 
좋은 지적 감사합니다 :)

**Reply by born-developer** [reply]
아무래도 주니어 개발자들에게 가장 유명하신 분이다 보니 (의도하지 않으셨더라도) 주니어들의 공부 방향에 영향을 많이 끼치실것 같아서, 90%의 경우 엔진이 자동으로 해주는 최적화를 열심히 공부하는 주니어가 생길까봐 정확하게 짚고 넘어가고 싶어서 적었습니다. 감사합니다.

**Reply by jojoldu** [reply]
네 ㅋㅋ 제가 학습 효율을 높이기 위해 사용하는 블로그가 어느 순간 말씀하신대로 주니어분들이 많이 참고하는 블로그가 되었더라구요 ㅋㅋ..
말씀해주신대로 그런 부분에서 주니어분들이 큰 그림 보다는 협소적인 최적화에 집중하시지 않도록 블로그 글 쓸때도 조심하겠습니다 :) 
좋은 말씀 감사합니다!

**Comment by dlfdnd96** [comment]
안녕하세요. 작성하신 내용 흥미롭게 잘 봤습니다.

저는 최근에 SQL 튜닝을 공부하고 있는데요. 공부한지 얼마 안 된 초보지만 본문을 보고 궁금한 점이 있어서 질문드립니다.
SQL 튜닝 책인 "오라클 성능 고도화 원리와 해법 2"에서는 Hash 조인의 사용기준을 다음과 같이 말하는데요.
> 1. 조인 컬럼에 적당한 인덱스가 없어 NL 조인이 비효율적일 때
> 2. 조인 컬럼에 인덱스가 있더라도 NL 조인 드라이빙 집합에서 Inner 쪽 집합으로의 조인 액세스량이 많아 Random 액세스 부하가 심할 때
> 3. 소트 머지 조인하기에는 두 테이블이 너무 커 소트 부하가 심할 때
> 4. 수행빈도가 낮고 쿼리 수행 시간이 오래 걸리는 대용량 테이블을 조인할 때
> 
> 수행시간이 짧으면서 수행빈도가 매우 높은 쿼리(=OLTP 쿼리 특징)를 Hash 조인으로 처리한다면? NL 조인에 사용되는 인덱스는 영구적으로 유지되면서 다양한 쿼리를 위해 공유 및 재사용되는 자료구조다. 반면, Hash 조인의 Hash 테이블은 단 하나의 쿼리를 위해 생성하고 조인이 끝나면 곧바로 소멸하는 자료구조다. 따라서 Hash 조인을 사용하면 CPU와 메모리 사용률을 크게 증가시키고 Latch(공유 리소스에서 사용되는 Lock 메커니즘) 경합이 발생해 시스템 동시성을 떨어뜨린다.

첫번째 질문은 4번 내용인데요. 본문 SQL은 수행빈도가 높아 보이는 것으로 예상되는데, 그렇다면 CPU와 메모리 사용량이 높아질텐데 혹시 다른 이유로 Hash 조인을 선택한 것인지 궁금합니다. 두번째 질문은 최종 SELECT-LIST 가 institutions 테이블의 컬럼인 것을 반영한 EXISTS 구문, 테이블 조인 순서 변경과 같은 JOIN 튜닝 이외의 것도 고민하셨다면 공유해 주실 수 있는지 궁금합니다.

**Reply by jojoldu** [reply]
@dlfdnd96 님 댓글 감사합니다 :) 
말씀하신대로 Hash Join이 CPU가 높은 작업이라서 모든 면에서 사용하기에는 어려움이 있는데요.
이번 개선 건의 경우 빈번하게 발생하는 경우가 아닌 일정 경우에 발생하는 건이며, 
슬로우 쿼리로 인한 시스템 병목이 더 문제라고 봤기 때문에 (이 쿼리를 Hash Join으로 변경한다고 발생하는 CPU 부하가 그리 크지 않았구요 ㅎㅎ)
개선적용을 하였습니다 :)

**Reply by dlfdnd96** [reply]
답변 감사합니다. 많은 도움 되었습니다~

**Comment by mystyle2006** [comment]
혹시 hint로 hash 조인을 유도해도 같은 결과가 나오는걸까요?
